mcumgr-core/src/main/java/io/runtime/mcumgr/managers/meta/StatisticsCollector.kt

```diff
@@ -0,0 +1,147 @@
+package io.runtime.mcumgr.managers.meta
+
+import io.runtime.mcumgr.McuMgrCallback
+import io.runtime.mcumgr.exception.McuMgrErrorException
+import io.runtime.mcumgr.exception.McuMgrException
+import io.runtime.mcumgr.managers.StatsManager
+import io.runtime.mcumgr.response.stat.McuMgrStatListResponse
+import io.runtime.mcumgr.response.stat.McuMgrStatResponse
+import java.util.concurrent.atomic.AtomicBoolean
+
+/**
+ * Result of statistic collections.
+ */
+sealed class StatCollectionResult {
+    data class Success(val statistics: Map<String, Map<String, Long>>): StatCollectionResult()
+    data class Cancelled(val statistics: Map<String, Map<String, Long>>): StatCollectionResult()
+    data class Failure(val t: Throwable): StatCollectionResult()
+}
+
+/**
+ * Callback for statistic collections.
+ */
+typealias StatCollectionCallback = (StatCollectionResult) -> Unit
+
+/**
+ * Non-blocking cancellable interface for cancelling an ongoing task.
+ */
+interface Cancellable {
+    fun cancel()
+}
+
+/**
+ * Collects stats from a device.
+ */
+class StatisticsCollector(private val statsManager: StatsManager) {
+
+    /**
+     * Collect stats from a single group by name.
+     */
+    fun collect(groupName: String, callback: StatCollectionCallback): Cancellable {
+        return StatCollection(statsManager, callback).start(listOf(groupName))
+    }
+
+    /**
+     * Collect from a list of statistic group names.
+     */
+    fun collectGroups(groupNames: List<String>, callback: StatCollectionCallback): Cancellable {
+        return StatCollection(statsManager, callback).start(groupNames)
+    }
+
+    /**
+     * List the stat group names from the device and collect each which intersects with the filter.
+     */
+    fun collectAll(filter: Set<String>? = null, callback: StatCollectionCallback): Cancellable {
```

Comment:
The coroutines is currently used in the unit tests, I can make it a testImplemetation

Reply:
The reason why the external API is not coroutines is to allow for people to use the traditional callbacks if they want (like we do...). Building coroutines support on top of this API would be very simple needed.

Reply:
If the plan is to make Coroutines first class citizen, then perhaps it would be better to use Coroutines in the library then provide an "extension" library that wraps the Coroutine versions to provide the callback implementation?

Reply:
I would love to have coroutines as a first-class citizen. The issue I see is that the current architecture for mcumgr does not include it whatsoever; so any coroutines addition to the current major version 0.x.x would simply be at a wrapping-api level. The crux of coroutine support would be at the transport interface.

I think we should plan a v1.x.x with coroutines built in all the way down to the transport interface and beyond, with a callback based api-wrapper (essentially the opposite of what we would do now). 

In the end, I think adding corountes would be great, just not so in this PR.